### ISSUE TITLE
Improve clarity of the error thrown by Dry::System::Loader#constant when a constant is not found

### ISSUE THREAD

**Issue by serodriguez68** [dry-rb/dry-system#147]
Hello!  

I want to suggest a minor but meaningful improvement to the clarity of the error that gets raised when `Dry::System::Loader` cannot load a constant.

**Disclaimer:**  I am a dry newbie so apologies in advance if what I'm proposing doesn't make sense in the broader scheme of `dry-system`.

```
As a developer
When I create a file that will be loaded through Dry::System
And the file I created has some typo in the name of the constant (or the path)
I expect Dry::System to point me directly to the file that cannot be loaded
```

## Examples

Let's say I create a file that is meant to be injected somewhere else and I have configured my app to auto-register files living on `lib`.  However, when I create the file, I make a typo on the constant name (or on the name of any `directory` on the path to the file).

```ruby
# lib/admin/articles/commands/create_article.rb
module Admin
  module Articles
    module Command # <= I missed an "s"
      class CreateArticle
         # ...
      end
    end
  end
end
```

When I load my application, I get a `NameError` that comes from `dry-inflector`. 

```
3: from .../gems/dry-system-0.17.0/lib/dry/system/loader.rb:52:in `call'
2: from .../gems/dry-system-0.17.0/lib/dry/system/loader.rb:67:in `constant'
1: from .../gems/dry-inflector-0.2.0/lib/dry/inflector.rb:90:in `constantize'
.../gems/dry-inflector-0.2.0/lib/dry/inflector.rb:90:in `const_get': uninitialized constant Admin::Articles::Commands (NameError)
```

However, the `NameError` doesn't directly point me to the source of the problem. Instead, I get pointed to `Admin::Articles::Commands`; but I may have many files under that namespace.

### Proposed Solution

Doing something like this in `Dry::System::Loader#constant`, can make the error reporting better:

```ruby
def constant
  stringified_const_name = inflector.camelize(path)
  inflector.constantize(stringified_const_name)
rescue NameError
  raise NameError, "uninitialized constant #{stringified_const_name}"
end
```

```
gems/dry-system-0.17.0/lib/dry/system/loader.rb:69:in `rescue in constant': uninitialized constant Admin::Articles::Commands::CreateArticle (NameError)
```

## Extra Notes
- I don't know if this should be implemented here or in `dry-inflector#constantize`, but I think it makes more sense in the context of `dry-system`.

cc @timriley

**Comment by timriley** [comment]
Thanks @serodriguez68, I think this is a great idea, and I agree that this is a better place to make a change than dry-inflector. I think we could actually go a little further:

- Make it a new exception class, e.g. `ConstantNotFound`
- If we can, include the name of the file in which we expected to find the constant defined
- As well as the new class name, like you noted

I wonder if we can also look at the `DidYouMean` gem for some inspiration re: error reporting.

**Comment by GuiHeurich** [comment]
hey there @timriley, @serodriguez68 . Is this something that still needs help with? I'd be happy to give it a go.

**Comment by timriley** [comment]
This is now fixed by #217. Thanks for bringing this helpful suggestion, @serodriguez68 :)
